### ISSUE TITLE
Avoiding instantiating linkify-it if it's disabled

### ISSUE THREAD

**Issue by fabiospampinato** [markdown-it/markdown-it#578]
I've taken some measurements, and in my system instantiating `linkify-it` takes most of the time needed for instantiating `markdown-it` itself:

<img width="771" alt="Screen Shot 2019-07-28 at 23 01 23" src="https://user-images.githubusercontent.com/1812093/62012888-23763d00-b18c-11e9-82fa-293fc19e166e.png">

`linkify-it` gets instantiated even if the `linkify` option is set to `false`. As a performance optimization `linkify-it` shouldn't get instantiated if the `likify` is set to `false`.

**Comment by puzrin** [comment]
Do you have examples when this is needed? Usualy one or several markdown-it instances are created and then reused everywhere. I can not imagine where saving of 2.8ms will be significant.

**Comment by fabiospampinato** [comment]
> Do you have examples when this is needed?

Yes, I use 2 instances of `markdown-it` in [notable](https://github.com/notable/notable), each with their own set of plugins enabled, one needs `linkify-it` and the other doesn't.

> I can not imagine where saving of 2.8ms will be significant.

Well I could split 1 minute in chunks of ~2.8ms, if none of them were significant then saving 1 minute wouldn't be significant either, which would be absurd.

I think of it this way:

- Each `markdown-it` instance that doesn't need `linkify-it` wastes ~3ms
- I have definitely users running the code in a computer at least 3x slower than mine, in the future I'll have users running the same code from mobile too, which may be 10x slower or more, so we are looking at ~9ms/~30ms under some scenarios.
- If I can find 10 smallish performance optimizations like this one I would save ~30ms from startup for me, ~90ms from startup for somebody with a slower computer and ~300ms or more from startup for somebody using a mobile device.

I think these figures significant enough to deserve addressing.

As kind of another data point to the matter it looks like Amazon found that every extra ~100ms costed them 1% in sales ([reference](https://www.gigaspaces.com/blog/amazon-found-every-100ms-of-latency-cost-them-1-in-sales/)), I'd bet that if they were wasting ~3ms here and there doing something totally unnecessary they'd would have optimized those a while ago.

**Comment by puzrin** [comment]
I don't understand, why do you run init every time instead of do it once? If user runs node.js every time, why do you ignore v8 startup time then, which is much longer?

I don't like spend time for premature optimizations "just for fun". So, having examples from real world is critical.

**Comment by fabiospampinato** [comment]
> I don't understand, why do you run init every time instead of do it once?

I create `markdown-it` instances only once, but if the user closes the window the instances are lost and will have to be recreated once the window is opened again.

> If user runs node.js every time, why do you ignore v8 startup time then, which is much longer?

I'm not ignoring it. Should I avoid shaving off ~100ms from startup (~3ms times a device 3x slower than mine times 10 optimizations like this one) because the process takes some amount of ms to startup anyway? If anything I believe I should try to compensate for electron/chromium/v8's startup overhead.

> I don't like spend time for premature optimizations "just for fun". So, having examples from real world is critical.

I can submit the PR if you'd like.

I'm not sure what kind of examples are you looking for, do you want me to list 10 possible performance optimizations I could make to my codebase each worth around 3ms? Do you think shaving 100ms+ from startup is insignificant?

One of the VSCode guys recently gave a [talk](https://www.youtube.com/watch?v=r0OeHRUCCb4) on optimizing the first second of VSCode startup time, does that look more like the kind of real life example you're looking for? Granted they probably aren't rendering any Markdown at startup, but I am.

**Comment by puzrin** [comment]
This is stable project with very conservative api update policy. Changes are done only when absolutely necessary and can't be avoided. I'd suggest you to use markdown-it several months and then, when you know all details, and still sure change needed - create issue again. Is it ok?

**Comment by fabiospampinato** [comment]
> This is stable project with very conservative api update policy.

If the issue is avoiding breaking compatibility with existing codebases the `linkify` property could be made a (cached) getter, so that even if somebody accesses the `linkify` property everything would still work as expected.

This optimization can require no API change whatsoever.

Was this the road blocker?

> I'd suggest you to use markdown-it several months and then, when you know all details, and still sure change needed - create issue again. Is it ok?

I don't understand the reasoning behind this, are you saying that you'd prefer not to merge in this optimization today, but merge it in in a few months?

I can't see a future where I'd think to myself "now these totally wasted 3ms (for each optimization) aren't worth optimizing".

**Comment by puzrin** [comment]
> I don't understand the reasoning behind this, are you saying that you'd prefer not to merge in this optimization today, but merge it in in a few months?

I have have time limits, can been spend for this project (including talks in tracker). Usually, new users have too much of energy and "fresh ideas", but that reduces significantly after some time. So, i'd like to not spend time and not change anything if that's possible.

When you can come here with words "i use markdown-it in real product, and it's the only package causing slowdown" - we can talk more seriously. Until that, it looks like talk about "faster code is better" - no possibility to participate, because busy with other things.

**Comment by fabiospampinato** [comment]
> Usually, new users have too much of energy and "fresh ideas", but that reduces significantly after some time.

Well I'm not surprised by this if no proposals are worth considering ever 😄.

> "i use markdown-it in real product, and it's the only package causing slowdown"

Yeah what kind of project would that be, come on. This is just like saying "I don't care about making my library start up 50%+ faster".

Anyway it's your library, your time (I'd take less to read the PR diff and the click the merge button honestly).

I'll set a reminder to myself a few months in the future to jump in in this thread again in case you changed your mind.

**Comment by puzrin** [comment]
> This is just like saying "I don't care about making my library start up 50%+ faster".

Yes. From my experience of code opts, talks about "faster/slower" are ineffective spend of time. I don't care about startup time if that's not critical chain.

Just create later a new issue, with example this package startup time is critical chain for your product, and we can move forward. Or we can do nothing if that's not critical chain.
